asf_heat_pump_affordability/notebooks/investigate_effect_of_room_number.py

```diff
@@ -0,0 +1,496 @@
+# 3. For each property type, compare the difference in pre- and post-1950 installation cost at each quantile between both model groups (a_models vs ar_models).
+#
+# Note: confidence intervals spanning zero indicate there is a chance the difference in coefficients for pre- and post-1950 archetype is zero and therefore that there may be no meaningful difference in cost of installation when splitting the archetype by pre- or post-1950 build year.
+
+archetypes4 = _data.archetype4.unique()
+X_y = _data[["archetype", "NUMBER_HABITABLE_ROOMS", "adjusted_cost"]]
+
+## Regex pattern to extract archetype names from summary table
+pattern = r"\[(.*?)\]"
+
+# ## Series 1: models with archetypes as only features
+
+# +
+a_model_pre_post_coefs = {
+    "quantile": [],
+    "archetype": [],
+    "pre_1950_coef": [],
+    "post_1950_coef": [],
+    "pre_1950_coef_lower": [],
+    "post_1950_coef_lower": [],
+    "pre_1950_coef_upper": [],
+    "post_1950_coef_upper": [],
+}
+
+for quantile in np.arange(0.1, 1, 0.1):
+    quantile = round(quantile, 1)
+    # Train model
+    archetype_model = sm.quantreg("adjusted_cost ~ archetype - 1", X_y).fit(quantile)
+
+    # Get summary table containing coefficients, p-values, and confidence intervals for the model
+    table = archetype_model.summary(alpha=math.sqrt(0.05)).tables[1]
+    _df = pd.read_csv(
+        StringIO(table.as_csv()), index_col=0
+    ).reset_index()  # convert statsmodels table to df
+    _df.columns = ["feature", "coef", "std err", "t", "P>|t|", "lower", "upper"]
+    _df["feature"] = _df["feature"].apply(lambda x: re.search(pattern, x).group(1))
+
+    # Create dict to generate df from for calculating coefficient difference and confidence intervals from
+    for a4 in archetypes4:
+        pre = _df[_df["feature"] == f"pre_1950_{a4}"]
+        post = _df[_df["feature"] == f"post_1950_{a4}"]
+
+        a_model_pre_post_coefs["quantile"].append(quantile)
+        a_model_pre_post_coefs["archetype"].append(a4)
+        a_model_pre_post_coefs["pre_1950_coef"].append(pre["coef"].values[0])
+        a_model_pre_post_coefs["post_1950_coef"].append(post["coef"].values[0])
+        a_model_pre_post_coefs["pre_1950_coef_lower"].append(pre["lower"].values[0])
+        a_model_pre_post_coefs["post_1950_coef_lower"].append(post["lower"].values[0])
+        a_model_pre_post_coefs["pre_1950_coef_upper"].append(pre["upper"].values[0])
+        a_model_pre_post_coefs["post_1950_coef_upper"].append(post["upper"].values[0])
+
+
+## NB:
+## Set alpha=np.sqrt(0.05) in `model.summary()` because we want to alter the confidence limits reported for the coefficients.
+## This is because we are interested in comparing the differences between 2 pairs of coefficients.
+## Leaving alpha=0.05 would reflect a confidence interval much larger than 95%, i.e. 99.75%,
+## because it would involve the comparison of extreme values at 2 * 95 percentile intervals instead of just 1.
+## We control for that by using sqrt 0.05.
```

Comment:
This reasoning is fine. The interval tells us something about the accuracy of the estimate compared to the population, the intuition being that for overlapping intervals we cannot conclude there is a 'true' difference. In this case we're not being that rigorous because we don't have a clear view what our population actually is.

In some senses we're interested in whether our results are internally consistent, without thinking too much about generalisability/external consistency. Effectively this sets a higher bar for internal consistency which I don't have a problem with.